충남대 FE 이호준 - 1단계 react query 적용

hojunleeWeb opened the pull request:
현재 선물 테마, 테마 상품 정보, 급상승 선물 랭킹에 대한 api 처리에 대한 react-query 적용이 완료되었으며,
로그인 및 주문 처리에 대한 react-query가 미완성입니다.

한번에 피드백 받는 것보다 이미 완성되있는 내용에 대해서 먼저 피드백을 받는것이 좋지 않을까 싶어 미완성인 상태에서 PR 드립니다.

hojun0010 committed:
feat : 프로젝트 0단계 - 클론

hojunleeWeb committed:
feat : 선물 테마 react-query 적용

hojunleeWeb committed:
feat : 급상승 선물랭킹 react-query 적용

hojunleeWeb committed:
feat : 상품 테마 -> 상품 정보에서 라벨 api 요청 react-qeury 적용

hojunleeWeb committed:
feat : 테마 상품 라벨 처리 로직이 get reqeust가 반환하는 값의 타입과 달라 스타일링이 적용되지 않는 문제 해결

hojunleeWeb committed:
feat : useQuery가 반환하는 isError를 사용 -> 기존의 error를 사용하는 컴포넌트나 훅들의 타입 및 변수 수정

hojunleeWeb committed:
feat : 선물 테마 상품 목록 react-qeury 적용

hojunleeWeb committed:
Merge branch 'hojunleeweb' into hojunleeweb

SeonJun-Hwang commented on src/components/Home/PresentRanking/Cotainer/PresentRankingContainer.tsx:
```diff
@@ -8,11 +8,18 @@ import {
+  const params = new URLSearchParams(search);
+  const rankType = params.get(PARAMS.rankType);
+  const targetType = params.get(PARAMS.targetType);
+  const typeUrls = `?targetType=${targetType}&rankType=${rankType}`;
```
0단계 PR에서는 세팅이라 넘겼던 부분이긴 한데
axios에는 해당 params 세팅을 도와주는 부분이 있습니다.
라이브러리를 사용하기전에는 해당 라이브러리에 공식문서가 있는지, 있다면 어떤 내용이 있는지 간단하게 API는 읽어보고 사용하시는게 좋아요

SeonJun-Hwang commented on src/components/Home/PresentRanking/Cotainer/useRankingItemFetch.ts:
```diff
@@ -0,0 +1,24 @@
+const BASE_URL = import.meta.env.VITE_API_BASE_URL;
+
+const getFetch = async (params: string): Promise<Good[]> => {
+  const res = await axios.get(BASE_URL + BASIC_ENDPOINT.ranking + params);
```
이부분도 0단계에서 넘겼던 부분이에요
url을 세팅함에 있어서 단순히 문자열로 잇게되면 사용성에 제한이 있어요
native 라이브러리중에 해당 역할을 도와주는 내용이 있으니 한번 알아보셔도 좋습니다.

SeonJun-Hwang commented on src/components/Home/PresentRanking/Cotainer/useRankingItemFetch.ts:
```diff
@@ -0,0 +1,24 @@
+    isError,
+    isLoading,
+  };
+};
```
useRankingItemFetch 라는 커스텀 훅에 비해 들어있는 내용이 너무 재사용성이 낮아요
재사용성이 높은 부분을 먼저 추출해내고 적용하는 과정이 필요해보입니다

SeonJun-Hwang commented on src/components/ThemesProduct/ThemesProductItem.tsx:
```diff
@@ -1,34 +1,31 @@
+            goods={(data?.pages ?? []).flatMap((page) => page.list)}
+            isError={isError}
+            isLoading={isLoading}
+          />
```
훅에서 불러온 내용이 컴포넌트 하나에 그대로 넘어가는데
굳이 외부에서 선언해서 넘기는 이유가 있나요?

hojunleeWeb replied:
말씀해주시기 전까지 컴포넌트에 바로 넘기는 방법을 몰라 찾아보았습니다만 
const hookData = useSomeHook();
// ...
<PresentProductList {...hookData} />
이러한 방법을 찾을 수 있었습니다. 이러한 방법을 사용했을 경우의 단점은 props로 전달되는 변수의 구성을 한눈에 알아볼 수 없어 가독성이 떨어지는 문제가 있을 수 있고, 더 적은 변수를 입력해 타자를 덜 친다는 장점이 있을 수 있겠지만 그정도로 많은 변수의 숫자가 아니라 생각해 가독성을 향상하고 컴포넌트로 넘어가는 것이 명확히 무엇인지 알아볼 수 있는 것이 더 좋지 않을까 생각합니다

SeonJun-Hwang replied:
질문을 잘못 이해하신거 같아요
props로 **어떻게** 넘기는지 가 아니라 **왜** 넘기는지에 대한 이유가 궁금합니다

SeonJun-Hwang reviewed (changes requested):
나머지 내용도 완성하신 후에 리뷰 요청 부탁드립니다.
반영에 앞서 일부 코멘트 참고 부탁드립니다.

hojunleeWeb committed:
feat : 피드백적용 - axios params 세팅 적용)

hojunleeWeb committed:
feat : 피드백 적용
포괄적인 기능을 정의한 apiQuery hook 생성,

hojunleeWeb committed:
feat : 피드백 적용 - useApiClient를 이용한 선물 테마 상품 라벨 생성

hojunleeWeb committed:
feat : 로그인 기능 react-query 적용

hojunleeWeb committed:
feat : 로그인 react-query 적용 수정

hojunleeWeb committed:
feat : 에러 메시지 처리를 위한 useApiQuery 수정

hojunleeWeb committed:
feat : 상품 주문 폼 - 상품 정보 get 요청 구현

hojunleeWeb committed:
feat : 주문 기능 react-query 적용

hojunleeWeb committed:
feat : 상품 주문 - 상품 정보 suspense 적용

hojunleeWeb committed:
fix : suspend 적용 삭제

hojunleeWeb committed:
Merge branch 'hojunleeweb' of https://github.com/hojunleeWeb/react-gift-product-detail into hojunleeweb

hojunleeWeb commented:
그리고 8월1일까지 2번 과제까지 완료하라는 공지문을 봐서 문의 드립니다.
피드백이 가능하신 시간대를 알려주시면 제가 그 시간대에 맞춰서 코드 작성 이후 코드 리뷰 알림을 드리면 좀 더 빠르게 처리가 가능할 것 같은데 피드백이 가능하신 시간대를 알려주시면 감사드리겠습니다

SeonJun-Hwang commented:
@hojunleeWeb  언제든 상관 없으니 코드 작성 이후 리뷰 요청만 주시면 됩니다 😄

hojunleeWeb commented:
혹시 다른 피드백은 아직 안해주신건지 아니면 혹시 1단게 중 빼먹은 부분이 있어 안해주신건지 잘 모르겠어서 그 점도 같이 알려주시면 감사드리겠습니다

SeonJun-Hwang commented on src/api/useApiQuery.ts:
```diff
@@ -0,0 +1,49 @@
+};
+type ApiQueryParams<T> = {
+  endpoint: string;
+  queryKey: any[];
```
any는 쉽게 쓰는 타입이 아닙니다.
any를 붙이는 곳에는 항상 그만한 이유가 필요합니다.

SeonJun-Hwang commented on src/api/useApiQuery.ts:
```diff
@@ -0,0 +1,49 @@
+
+  const nullNotData = data ?? null;
+  return { nullNotData, isError, isLoading };
+};
```
훅 자체가 useQuery를 쓰기만 한 것일뿐 오히려 기능을 크게 억제하는거 같아요
이 커스텀 훅 보다 useQuery 자체만 쓰는것이 나아보입니다.

SeonJun-Hwang commented on src/api/useApiQuery.ts:
```diff
@@ -0,0 +1,49 @@
+  const { data, isError, isLoading } = useQuery<T, AxiosError<ApiErrorResponse>>({
+    queryKey,
+    queryFn: async ({ signal }) => {
+      const response = await axios.get(BASE_URL + endpoint, {
```
이부분도 코멘트가 반영이 안된거 같아요
이런 형태로 사용하게 되면 BASE_URL 뒤에 `/`가 있거나 없어야 하는 내용이 강제가 되는 그림이 될거같아요

SeonJun-Hwang commented on src/components/Login/useLoginQuery.ts:
```diff
@@ -0,0 +1,50 @@
+  const response = await axios.post(BASE_URL + BASIC_ENDPOINT.login, body);
+  return response.data;
+};
+export const useLoginQuery = () => {
```
query 파일 내에 axios 호출부 부터 있게 되면
차후에 여러곳에서 호출할 수 있는 API 같은 경우에 API를 놓는 위치가 애매해질듯 한데 API끼리 모아놓는건 어떨까요

SeonJun-Hwang commented on src/components/Order/Container/useOrderForm.ts:
```diff
@@ -71,44 +48,22 @@ export const useOrderForm = () => {
+      }),
     };
-    postOrder();
+    postOrder(orderBody);
```
외부에서 데이터를 정제해서 보내는 것과
커스텀 훅 내부에서 API 데이터 포맷으로 맞춰서 보내는 것이 어떤 차이가 있을까요

SeonJun-Hwang commented on src/components/Order/Container/usePostOrderForm.ts:
```diff
@@ -0,0 +1,74 @@
+  statusCode: number;
+};
+
+const BASE_URL = import.meta.env.VITE_API_BASE_URL;
```
여러군데서 사용되는 것 같은데 axios 기능 중에 관련된 걸 편하게 해줄 수 있는 방법은 없을까요?

SeonJun-Hwang commented on src/components/ThemesProduct/ThemesProductionLabel.tsx:
```diff
@@ -0,0 +1,23 @@
+  if (isLabelLoading) {
+    return <div>로딩 중</div>;
+  } else if (labelError) {
+    console.log(labelError);
```
PR 요청에는 디버그를 위한 console은 없는게 좋아요
자주 놓칠 수 있는 부분인데 이런걸 레포에 push 하기전에 확인 할 수 있는건 없을까요?

SeonJun-Hwang commented on src/hooks/usePresentThemeFetch.ts:
```diff
@@ -0,0 +1,25 @@
+  const res = await axios.get(BASE_URL + BASIC_ENDPOINT.theme);
+  const data = res.data;
+  return data;
+};
```
여기도 동일하게 전체적으로 코드가 반복되고 있어요

SeonJun-Hwang commented on src/hooks/useThemesProductItem.ts:
```diff
@@ -0,0 +1,51 @@
+  });
+  const data = res.data.data;
+  return data;
+};
```
여기도 동일하게 API 호출부가 반복되네요

SeonJun-Hwang reviewed (changes requested):
죄송하지만 어떤 의미로 코멘트를 남기시는지 이해가 잘 안됩니다.

코드리뷰는 상호 존중을 기반으로 하고 진행하는것이 기본이라고 생각합니다.
어떤 생각을 가지고 개발했는지를 듣기위함이고, 서로 의견 교환을 통해 서로 성장하는것이 목표입니다.

지금 호준님은 학교 수업으로 질문을 주시는 과정이고, 단순히 리뷰보다 Approve만 바라고 저에게 PR 요청을 주시는 것으로밖에 보이지 않습니다.

1차 코멘트로 일부만 구현하고 피드백을 나머지 코드에 적용하겠다 하셔서 기존 코드만으로도 전부 리뷰를 해드렸습니다.
2차 수정까지 코멘트에 어떤 목적으로 작성했는지에 대한 대답은 없었고, 충분히 리뷰가 반영되지 않았습니다.

그리고 3차로 저에게 왜 Approve를 주지 않는지를 물어보신다면, 2차도 반영이 안됬고, 기존 미완성 기능이 구현이 되었는지도 PR에 작성이 안되어 있는데, 보는 사람 입장으로서는 리뷰 코멘트를 읽어보긴 한건가 하는 의심이 더 강합니다.

단순 Approve만 바라신다면 3차까지 어떤 코멘트 없이 넘겨드리겠습니다. 무의미한 코드리뷰는 하고싶지 않습니다

hojunleeWeb committed:
refactor : 파일 위치 변경

hojunleeWeb committed:
refactor : 선물 테마 관련 코드 수정
- 피드백 반영 : useApiClient는 자체가 기능을 크게 억제함으로 각 컴포넌트의 훅 자체에서 useQuery를 사용하게 하고 그 대신 재사용성을 위해 공동으로 사용하는 queryFn 함수를 제네릭 타입을 이용해 정의

hojunleeWeb committed:
refacor : getFetch 함수가 params를 받아와 axios 요청에 사용할 수 있도록 재정의

hojunleeWeb committed:
refactor : useApiClient를 사용하지 않고 useSuspenseQuery를 이용해 급상승 랭킹 아이템을 구현
- useSuspenseQuery를 이용한 이유 : data에 값이 존재하는 것이 보장되기 때문에 Optional Chaining할 필요가 없고, Suspense를 활용한 isLoading 처리를 간편하게 함

hojunleeWeb committed:
refactor : useRankingItem 훅이 SuspenseQuery를 사용함에 따라 error 처리, 로딩 처리를 위한 ErrorBoundary, Suspense fallback 사용
refactor : PresentRankingItem 수정
- props로 data를 전달받지 않고 컴포넌트 내부에서 data를 반환하는 훅을 사용하도록 수정
-추후 RankingItem을 재사용하되 api 요청 url이 다른 경우에는 파라미터를 통한 구분을 통해 다른 api 훅을 사용하도록 할 예정

hojunleeWeb committed:
refacor : getBasicFetch 파일 이름 수정 - getFetch

hojunleeWeb committed:
refactor(useLoginQuery) : api 처리 부분은 api 폴더로 이동(postFetch), 제네릭 타입을 사용해 추후 post 요청하는 경우를 고려하여 정의

hojunleeWeb committed:
refactor : 상품 주문(상품 정보) - useQuery를 사용, getFetch를 재사용해 get 요청 처리

hojunleeWeb committed:
refactor (usePostOrderForm) : postFetch를 재사용하여 post 요청, headers를 파라미터로 전달하기 위해 postFetch 재정의

hojunleeWeb committed:
fix : 상품 주문(상품 정보) get 요청 - 잘못된 url 수정

hojunleeWeb committed:
refactor(useOrderForm) : 상품 주문(상품 정보) get 요청의 response 형태에 따라 type 재정의, selectedProduct 재정의

hojunleeWeb committed:
refactor : useQuery, getFetch를 이용한 선물 테마 상품 라벨 생성

hojunleeWeb committed:
fix : 테마 상품 라벨 get response 형태에 따라 ThemeLabel 타입 재정의, 리턴할 변수 수정

hojunleeWeb committed:
refactor(useOrderForm) : 상위 컴포넌트인 useOrderFOrm에서 받아온 변수를 재정의하지 않고 useGetOrderForm의 리턴하는 변수를 수정해 .data로 접근하지 않도록 수정

hojunleeWeb committed:
fix(ThemesProductionLabel) : useThemesProductLabel 수정에 따른 변수 할당 수정

hojunleeWeb committed:
refactor(PresentProductList) : ui 생성이라는 역할을 관리한다는 관점으로 props로 data를 전달받아 ui를 생성하도록 재정의

hojunleeWeb committed:
bug : 급상승 선물 랭킹을 생성하는데 초기값인 data가 undefined가 된 것에 반해 ErrorBoundary 등이 작동하지 않아 slice가 실행되어 오류를 발생

hojunleeWeb committed:
fix : PresentRankingList를 다시 props를 통해 상위 컴포넌트에서 data를 받아 ui를 생성하는 것으로 변경해 해결, useRankingItem에 useSuspenseQuery가 아닌 useQuery로 수정

hojunleeWeb committed:
fix : 선물 테마 상품 get response 타입에 맞추어 FetchDataType type을 재정의해 오류 해결

hojunleeWeb committed:
fix : PresentProductList의 showRankingNumber의 값에 따라 return 되는 로직이 다른데 이를 구분하지 않아 생성되는 컴포넌트의 숫자가 제한되어 테마 상품의 수가 제한되는 문제 해결

hojunleeWeb committed:
refactor : useApiQuery 삭제

hojunleeWeb committed:
fix : 로그인 이전 상품 클릭 -> 로그인 화면 -> 주문 화면으로 넘어가는 과정에서 잘못된 url 설정으로 not foun page로 넘어가는 문제 해결

hojunleeWeb commented:
- [x] 본인만의 기준으로 일관된 코드를 작성해주세요.
- [x] 기능 단위로 나누어 커밋을 해주세요. 
- [x] 기존에 작성했던 API를 React Query를 이용해서 리팩터링 해요. (GET, POST 모두)

1. 말씀해주신 기본 url을 설정하는 방법을 찾아 api/baseUrl.ts에서 axios.create를 하는 방법으로 구현하였습니다.
2. api 요청을 하는 대부분의 코드에서 공통적으로 사용하는 get, post 요청에 대한 코드를 제네릭 타입을 이용하여 재사용성을 높일 수 있도록 정의하였습니다.
3. get 요청을 하는데 추가적으로 필요한 파라미터 등을 받을 수 있도록 get 요청에 대한 코드인 getFetch.ts 함수에 props를 통해 값을 전달하여 get 요청에 포함시킬 수 있도록 하였습니다.
4. 현재 해결하지 못한 문제로는 axios.get(`${BASIC_ENDPOINT.theme}/${themeId}/info`, {}), 와 같이 여러 변수로 이루어진 url로의 전송에서 이를 문자열로 연결하는것이 아닌 방법을 찾는데 실패하였습니다. 혹시 이에 대해 조언해주실 내용이 있으실까요?

피드백 해주신 내용을 읽어봤는데 그러한 피드백을 처음 받아 말씀해주신 내용을 제대로 이해하였는지 의아한 부분이 있습니다. 혹시 말씀해주신 내용이 위와 같은 내용이 맞을까요?
